idx/memory/tag_query.go

```diff
@@ -0,0 +1,411 @@
+				resultSet[id] = struct{}{}
+			}
+		}
+		return startId, resultSet, nil
+	}
+
+	re, err := regexp.Compile(q.match[startId].value)
+	if err != nil {
+		return 0, nil, errInvalidQuery
+	}
+
+	for v, ids := range index[q.match[startId].key] {
+		if !re.MatchString(v) {
+			continue
+		}
+
+		for id := range ids {
+			resultSet[id] = struct{}{}
+		}
+	}
+	return startId, resultSet, nil
+}
+
+func (q *TagQuery) getInitialByEqual(index TagIndex) (int, TagIDs) {
+	resultSet := make(TagIDs)
+	lowestCount := math.MaxUint32
+	startId := 0
+
+	// choose key-value combo that has the smallest number of series IDs because
+	// we want to start with the smallest possible resultSet.
+	// the smaller the resultSet the less time will have to be spent to further
+	// filter its values in the following expressions.
+	for i := range q.equal {
+		l := len(index[q.equal[i].key][q.equal[i].value])
+		if l < lowestCount {
+			lowestCount = l
+			startId = i
+		}
+	}
+
+	// copy the map, because we'll later delete items from it
+	for k, v := range index[q.equal[startId].key][q.equal[startId].value] {
+		resultSet[k] = v
+	}
+
+	return startId, resultSet
+}
+
+// filterByEqual filters a list of metric ids by the given expressions. if "not"
+// is true it will remove all ids of which at least one expression is equal to
+// one of its tags. if "not" is false the functionality is inverted, so it
+// removes all the ids where no tag is equal to at least one of the expressions.
+//
+// expressions: the list of key & value pairs
+// skipEqual:   an integer specifying an expression index that should be skipped
+// resultSet:   list of series IDs that should be filtered
+// index:       the tag index based on which to evaluate the conditions
+// not:         whether the resultSet shall be filtered by a = or != condition
+//
+func (q *TagQuery) filterByEqual(expressions []kv, skipEqual int, resultSet TagIDs, index TagIndex, not bool) {
+	for i, e := range expressions {
+		if i == skipEqual {
+			continue
+		}
+
+		indexIds := index[e.key][e.value]
+
+		// shortcut if key=value combo does not exist at all
+		if len(indexIds) == 0 && !not {
+			for id := range resultSet {
+				delete(resultSet, id)
+			}
+		}
+
+		for id := range resultSet {
+			if _, ok := indexIds[id]; ok == not {
+				delete(resultSet, id)
+			}
+		}
+	}
+}
+
+// filterByMatch filters a list of metric ids by the given expressions. it is
+// assumed that the given expressions are all based on regular expressions.
+//
+// expressions: the list of key & value pairs
+// skipMatch:   an integer specifying an expression index that should be skipped
+// resultSet:   list of series IDs that should be filtered
+// byId:        ID keyed index of metric definitions, used to lookup the tags of IDs
+// not:         whether the resultSet shall be filtered by the =~ or !=~ condition
+//
+func (q *TagQuery) filterByMatch(expressions []kv, skipMatch int, resultSet TagIDs, byId map[string]*idx.Archive, not bool) error {
```

Comment:
some questions about this function:
1)  correct me if i'm wrong but the key must be an exact match right, so there can be only 0 or 1 value we have to pattern match per MetricDefinition (if true, please add comment to the function). it seems a bit wasteful to use notMatchingTags when the key doesn't even match, which can easily/cheaply be checked up front .
(I wonder how `strings.HasPrefix(tag, e.key + "="` would perform, it's all the len checks + check for the exact key match followed by equal sign, all in one)

2) this is an interesting approach (let's call A), esp. when compared to the approach taken elsewhere to work with the data from `index[e.key] ` (let's call B). I guess the tradeoff if this:

* A: each MD may contain several tags, and for the key we're looking for, they might have the same tag values. so this search is `O(number-of-mds * number-of-tags-per-md)` though the caches help alleviate this, and the number-of-tags-per-md is only a minor factor since we should be able to quickly iterate or break out of the loop.

* B: `index[e.key]` may contain several `TagValue`'s, each entry is a value pointing to several MetricID's. searching through this would be `O(number-of-tag-values-across-all-mds * number-of-mds-per-value)` . we could also add a caching system here though.

I think whether one approach is better than the other depends on whether we're able to get a very-narrowed-down list of results as input (then A is better) or not (then B may better because we can only process each value once)

anyway at this point i don't think we can predict much and probably not worth to put more effort into it.

Reply:
My 2 cents:
1. The cache should definitely only be used to avoid checking the regex. I thought that it was doing that, but I definitely skimmed that code incorrectly. `HasPrefix` was my first thought, but it's worst case for prefixMatching is slightly worse than this approach. Also, would only want to build `e.key +"="` once. I would guess that the performance is going to be near identical and/or washed out by the high cost of the regex, so the cleanest approach wins.

2. I had a similar thought when initially reviewing the code, but I appreciate the filtering approach (assuming that the initial filter(s) are effective). We will end up with some tags that have cardinality in the hundreds of thousands, so an approach that wouldn't need to scan them all if another cheaper filter is available is appealing. Of course, with additional complexity, this function could probably determine that by doing a lookup on the target key, determining it's cardinality and opting between approach A or B based on the size of the resultSet (recognizing that with approach B, no caching would be available since each tag value would likely only be in there once).

Reply:
you're right. fixed that here: https://github.com/grafana/metrictank/pull/729/commits/34f5323dcda5ba4f88e9500155c8272262c9e4f9
so the cache only caches the string that's matched with the regex now, if the key is different than what we're looking for the cache isn't even checked.